sunscreen/src/compiler.rs

```diff
@@ -398,37 +458,131 @@ impl<T, B> GenericCompiler<T, B> {
+            // programs should only be present when params are available.
+            let params = params.expect("no params; please file a bug!").clone();
+            // Note: this is currently unsupported because determining the exact arbitrary
+            // plaintext modulus from the dynamic length of a shared ZKP input in general is not
+            // possible (as x.ilog2() is not injective). We may support this in the future.
+            if !params.plain_modulus.is_power_of_two() {
+                return Err(Error::unsupported(
+                    "Plaintext modulus must be a power of two for ZKP programs with #[shared] arguments.",
+                ));
+            }
+            let result = prog.build(params.plain_modulus)?;
+            let result = zkp::compile(&result);
+            let metadata = ZkpProgramMetadata {
+                params: Some(params),
+                signature: prog.signature(),
+            };
+            let compiled_program = CompiledZkpProgram {
+                zkp_program_fn: result,
+                metadata,
+            };
+            Ok((prog.name().to_owned(), compiled_program))
+        });
         let zkp_programs = zkp_data
             .zkp_program_fns
             .iter()
             .map(|prog| {
-                let result = prog.build()?;
+                let result = prog.build(())?;
                 let result = zkp::compile(&result);
+                let metadata = ZkpProgramMetadata {
+                    params: None,
+                    signature: prog.signature(),
+                };
+                let compiled_program = CompiledZkpProgram {
+                    zkp_program_fn: result,
+                    metadata,
+                };
 
-                Ok((prog.name().to_owned(), result))
+                Ok((prog.name().to_owned(), compiled_program))
             })
+            .chain(shared_zkp_programs)
             .collect::<Result<HashMap<_, _>>>()?;
 
         Ok(zkp_programs)
     }
+}
 
-    fn compile_internal(self) -> Result<Application<T>> {
-        Application::new(HashMap::new(), self.compile_zkp()?)
+impl<B> ZkpCompiler<B>
+where
+    B: FieldSpec,
+{
+    /**
+     * Add the given FHE program for compilation.
+     */
+    pub fn fhe_program<F>(self, fhe_program_fn: F) -> FheZkpCompiler<B>
+    where
+        F: FheProgramFn + 'static,
+    {
+        let mut fhe_data = FheCompilerData::default();
+        fhe_data.fhe_program_fns.push(Box::new(fhe_program_fn));
+
+        FheZkpCompiler {
+            data: CompilerData::new_fhe_zkp(fhe_data, self.data.unwrap_zkp()),
+            _phantom: PhantomData,
+        }
+    }
+
+    /// Add the given ZKP program for compilation.
+    ///
+    /// Note that this method will not accept "shared" ZKP programs, which have inputs shared from
+    /// FHE programs. You must first call `fhe_program` to add an FHE program to unlock this
```

Comment:
`fhe_program` is always callable, so the note doesn't apply there. And yes the verbiage is correct, you need to have called `fhe_program`, and then you'll have a compiler whose `.zkp_program` accepts shared ZKPs. Otherwise you'll get a Rust-compile-time error.